Fixed #25734 -- Made GDALBand min and max properties use GDALComputeRasterStatistics.

yellowcap opened the pull request:
Properties now use the GDALComputeRasterStatistics function to retrieve values.

sir-sigurd commented:
It looks like statistics could be changed after calling `data`, but cached statics will be returned.

yellowcap commented:
Thanks, good point @sir-sigurd. I updated the PR to reset the statistics cache on call of the `data` method and setter of `nodata_value` (the statistics might also change when the nodata_value is changed).

timgraham commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -19,6 +19,14 @@ def __init__(self, source, index):
 
+    def _flush(self):
+        """
+        Calls the flush method on the Band's parent raster and
```
For new docstrings, we are using PEP 0257 verb style ("Call.... and clear...")
Below, "Return..."

sir-sigurd commented:
Probably it worth to check that the statistics cache is flushed properly in tests.

yellowcap commented:
Added tests for flushing, updated the functionality to work on empty rasters (where all values are nodata), and added documentation for the new methods/attributes.

timgraham commented:
It seems we now leak a file, `tests/gis_tests/data/rasters/raster.tif.aux.xml`, when running the tests.

Here are the edits I made before discovering that issue: http://dpaste.com/3Q2EK22 (careful of trailing whitespace in docs)

Feel free to enhance the release note if you see fit.

yellowcap commented:
After digging deeper into the gdal functions, I realized that for most formats, gdal persists statistics into an auxiliary files called Persistent Auxiliary Metadata (PAM) files.

I think that its better to use this than to cache the stats in the python objects. So I changed the `statistics` funcion to selectively use `GDALGetRasterStatistics` or `GDALComputeRasterStatistics` depending on the whether the cache in the aux files needs updating. Both functions create the aux file if it does not exist.

I added a note about that in the documentation and updated the tests to check for the presence of the files and the remove the aux file on `tearDown`.

timgraham commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -309,15 +309,33 @@ def setUp(self):
         rs = GDALRaster(self.rs_path)
         self.band = rs.bands[0]
 
+    def tearDown(self):
```
I think usually we use a try/finally block if this is only applicable to 1 test.

timgraham commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -47,19 +56,80 @@ def pixel_count(self):
+        The return value is a tuple with the following structure:
+        (minimum, maximum, mean, standard deviation).
+
+        If approximate=True, the statistics may be computed based overviews, or
```
"computed based overviews" is there a missing word? e.g. "computed based on overviews"?
I'm not sure of the exact meaning, but probably chop the comma before "or".

yellowcap replied:
Indeed, the word "on" was missing here. Rasters files can have "overviews" (basically pyramids of lower resolution version of the same raster) and/or internal tiling, which gdal can leverage to compute approximate statistics faster.

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1413,6 +1413,36 @@ blue.
+        Note that if a persistent cache value is found, that value is returned.
+        For raster formats using Persistent Auxiliary Metadata (PAM) services,
+        the statistics might be cached in an auxiliary file. In some cases this
+        metadata might be out of sync with the pixel values or store
```
maybe "or cause values from a previous call to be returned which don't reflect the value of the approximate argument" (assuming this problem can happen both ways)?

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1413,6 +1413,36 @@ blue.
+        computed from the data directly, and the cache will be updated with the
+        result.
+
+        Note that if a persistent cache value is found, that value is returned.
```
I try to avoid phrases like "Note that" -- it doesn't seem to add much.

yellowcap commented:
Integrated latest changes: improved wording and try/finally block for aux file removal.

timgraham commented:
merged in 8f5904560a7ab51d480410ca5228df1a28ec2295, thanks!
